### PR TITLE
config: v2 non-pubsub transport API fatal-by-default.

### PR DESCRIPTION
**Pull request by htuch**
This is a followup to #14223, covering remaining uses of the
transport_api_version field.

Risk level: High (this will break anyone who is still using v2 and has
   not enabled CLI or runtime override)
Testing: Various tests updated, some exemplar tests added to
  server_test.
Release Notes: Same as #13950.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/config/utility.h [hunk 1]
```diff
@@ -182,6 +183,34 @@ class Utility {
       const Upstream::ClusterManager::ClusterSet& primary_clusters,
       const envoy::config::core::v3::ApiConfigSource& api_config_source);
 
+  /**
+   * Access transport_api_version field in ApiConfigSource, while validating version
+   * compatibility.
+   * @param api_config_source the config source to extract transport API version from.
+   * @return envoy::config::core::v3::ApiVersion transport API version
+   * @throws DeprecateDMajorVersion when the transport version is disabled.
```
Comment (dio) [inline source/common/config/utility.h:191]: Super small nit s/DeprecateDMajorVersion/DeprecatedMajorVersionException

File: source/common/config/utility.h [hunk 2]
```diff
@@ -182,6 +183,34 @@ class Utility {
...
+   */
+  template <class Proto>
+  static envoy::config::core::v3::ApiVersion
+  getAndCheckTransportVersion(const Proto& api_config_source) {
+    const auto transport_api_version = api_config_source.transport_api_version();
+    if (transport_api_version == envoy::config::core::v3::ApiVersion::AUTO ||
```
Comment (dio) [inline source/common/config/utility.h:197]: Not sure if people will argue that "AUTO" should automatically use the currently "stable" version?

Comment (htuch) [reply source/common/config/utility.h:197]: Yeah, we agreed to deprecate `AUTO` in #14233.

**Review by mattklein123 — APPROVED** [review]
Nice!

**Pull request merged**